CMakeLists.txt

```
@@ -20,30 +23,74 @@ include(${_project_options_SOURCE_DIR}/Index.cmake)
+
+# Set defaults for options enabled with the project_options project
+# and the GlobalOptions utility below
+#
+# Note: These defaults represent those suggested by Jason Turner in
+# C++ Best Practices
+set(ENABLE_CPPCHECK
+    ON
+    CACHE BOOL "Enable CppCheck analysis")
+set(ENABLE_CLANG_TIDY
+    ON
+    CACHE BOOL "Enable clang-tidy analysis")
+set(ENABLE_CACHE
+    ON
+    CACHE BOOL "Enable ccache")
+set(ENABLE_COVERAGE
+    ON
+    CACHE BOOL "Enable coverage testing")
+set(ENABLE_CONAN
+    ON
+    CACHE BOOL "Enable conan for dependencies")
+set(ENABLE_SANITIZER_ADDRESS
+    ON
+    CACHE BOOL "Enable address sanitizer for runtime memory checking")
+set(ENABLE_SANITIZER_UNDEFINED_BEHAVIOR
+    ON
+    CACHE BOOL "Enable undefined behavior sanitizer for runtime checking")
+set(WARNINGS_AS_ERRORS
+    ON
+    CACHE BOOL "Enable undefined behavior sanitizer for runtime checking")
+
+# Add global CMake options
+include(${_project_options_SOURCE_DIR}/src/GlobalOptions.cmake)
+
 # Initialize project_options variable related to this project
 # This overwrites `project_options` and sets `project_warnings`
 # uncomment the options to enable them:
 project_options(
-      ENABLE_CACHE
-      # WARNINGS_AS_ERRORS
-      ENABLE_CPPCHECK
-      ENABLE_CLANG_TIDY
-      ENABLE_CONAN
-      ENABLE_COVERAGE
-      # ENABLE_IPO
-      # ENABLE_INCLUDE_WHAT_YOU_USE
-      # ENABLE_PCH
-      # PCH_HEADERS
-      # ENABLE_DOXYGEN
-      # ENABLE_USER_LINKER
-      # ENABLE_BUILD_WITH_TIME_TRACE
-      # ENABLE_UNITY
-      # ENABLE_SANITIZER_ADDRESS
-      # ENABLE_SANITIZER_LEAK
-      # ENABLE_SANITIZER_UNDEFINED_BEHAVIOR
-      # ENABLE_SANITIZER_THREAD
-      # ENABLE_SANITIZER_MEMORY
-      # CONAN_OPTIONS
+  ${ENABLE_CACHE} # ccache on Unix, to avoid rebuilding the same file with clean builds
+  ${WARNINGS_AS_ERRORS} # force warnings to be dealt with, by making them hard errors
+  ${ENABLE_CPPCHECK} # cppcheck is a standalone static analysis tool with helpful checks
+  ${ENABLE_CLANG_TIDY} # clang-tidy is a static analysis tool that's part of the llvm toolchain
+  ${ENABLE_CONAN} # package manager for getting dependencies with nice cmake integration
+  ${ENABLE_COVERAGE} # analyze and report coverage of code lines
+  ${ENABLE_IPO} # inter-procedural-optimization, aka whole program optimization aka link-time-optimization (LTO)
+  ${ENABLE_INCLUDE_WHAT_YOU_USE} # analysis to make sure #includes are handled properly
+  ${ENABLE_PCH} # pre-compiled headers. Can speed up compilation in some cases
+  # PCH_HEADERS a_list_of.hpp headers_to_use.hpp  # This is a list of headers to pre-compile
+  ${ENABLE_DOXYGEN} # build documentation with the doxygen tool
+  ${ENABLE_USER_LINKER}
+  ${ENABLE_BUILD_WITH_TIME_TRACE} # generates report of where compile-time is spent
+  ${ENABLE_UNITY} # merge C++ files into larger C++ files, can speed up compilation sometimes
+  ${ENABLE_SANITIZER_ADDRESS} # make memory errors into hard runtime errors (windows/linux/macos)
```

This is all it takes to enable ASAN by default. There is no need for adding all these global options. 

```suggestion
  ENABLE_SANITIZER_ADDRESS # make memory errors into hard runtime errors (windows/linux/macos)
```

---

That does not enable asan by default. Instead, it hardcodes the ENABLE_SANITIZER_ADDRESS and gives the user a button in the cmake-guis that does not do anything.

The work I have prototyped in here is so that a user of one of the GUIs for changing CMake settings has buttons that work how you would expect them to work. All the buttons do something and properly reflect the state of the system.

I just did 3 different clean (full delete of the build folder) builds with your suggestions to verify that what I did was necessary to make both worlds work, but I may still be missing something.

---

Either way, I'm still prototyping here, which is why I made it a draft PR. I think once I come up with something that I think works, it would be wise for us to have a zoom chat or something and discuss the best ways for it to work within the framework you've mad.e

---

> it would be wise for us to have a zoom chat or something and discuss the best ways for it to work within the framework you've made

Sure. That's a good idea! I am available for Zoom chats in the evenings (PST time).  Send me an invitation to my email, and I will put it on my calendar.

> hardcodes

That's why I call this a declarative build system that does not depend on external CLI flags and environment variables. For sanitizers, it is hard to decide in advance (that's why I automated it for tests as an example). But for things like using Conan, this decision can be definitely made and declared in the build.

_Side note rant_: 

_I cannot tell how many times I could not use a C++ library because of how broken their build script has been. A gazillion number of flags that I need to configure so I can use a library. This is in my opinion one of the biggest flaws of the current ecosystem. The C++ code even does not find a chance to run. People already give up on compiling. Either they switch to dynamic languages that do not require building or go to declarative style languages like cargo rust._

>  gives the user a button in the cmake-guis that does not do anything.

That's an issue we need to fix.  I overlooked the GUI aspect of the `GlobalOptions.cmake` file. I think if that is the case we can remove this file, and just let the user define the dynamic options they want to change on the fly. It seems the options file creates more confusion than the things it provides.